### ISSUE TITLE
Image removed from ceph but not in the database

### ISSUE THREAD

**Issue by naved001** [CCI-MOC/ims#122]
I was importing and removing images from the bmi db and ceph. not sure what went wrong, but the image was removed from ceph and not the database, leaving the db in an inconsistent state.

**Comment by apoorvemohan** [comment]
is this reproducible?

**Comment by chemistry-sourabh** [comment]
@naved001 did you inspect the logs ?

**Comment by pgrosu** [comment]
Shouldn't the `bmi import` command create a clone in the database?  Somehow it thinks it is there, but I do not see it:

```
(ims) [root@rhel-tgt-base ims]# rbd ls | grep bmi-test-image
bmi-test-image
(ims) [root@rhel-tgt-base ims]# bmi import bmi_infra bmi-test-image
snapshot Already Exists
(ims) [root@rhel-tgt-base ims]# bmi db ls
+----+-------------------+-----------+-------+--------+----------+--------+
| Id |        Name       |  Project  |  Ceph | Public | Snapshot | Parent |
+----+-------------------+-----------+-------+--------+----------+--------+
| 1  |     naved-img1    | bmi_infra | 5img1 | False  |  False   |        |
| 3  | centos7-radostyle | bmi_infra | 5img3 | False  |  False   |        |
| 4  |   centos7-naved   | bmi_infra | 5img4 | False  |  False   |        |
| 5  |     paul-image    | bmi_infra | 5img5 | False  |  False   |        |
+----+-------------------+-----------+-------+--------+----------+--------+
(ims) [root@rhel-tgt-base ims]#
```

**Comment by apoorvemohan** [comment]
@naved001 Were you able to do the sanity check that we discussed?

**Comment by chemistry-sourabh** [comment]
@pgrosu sorry for the late reply. The issue is that bmi import first create a snapshot with the name "snapshot" for the image and then does a clone. I am guessing that your image already has such a snapshot. If you delete it, then import should work.

**Comment by chemistry-sourabh** [comment]
@naved001 if there is an officially sanity check then make it into a document so that other people can refer when testing there stuff. Also if this check can be made into test cases then it is even better.

**Comment by naved001** [comment]
@chemistry-sourabh there's no official sanity check AFAIK. Apoorve suggested that sometimes the image may stay in ceph but it's name might get removed. Basically he wanted me to double check in ceph. @apoorvemohan correct me if I'm wrong.

**Comment by pgrosu** [comment]
Thanks @chemistry-sourabh and no worries :)  The actual name is "snapshot" or the one that appears under the `Ceph` column?  Could you point me to the program flow that performs these specific steps?

**Comment by chemistry-sourabh** [comment]
No, import actually failed. Do a rbd ls -l and you should see <image>@snapshot for your image. Delete it using rbd snap rm then try importing again. If it is protected you will need to unprotect it.

check for import method in operations.py

**Comment by apoorvemohan** [comment]
@naved001 You are correct. Were you able to reproduce it?

**Comment by naved001** [comment]
@apoorvemohan Nope, I still haven't had the time to poke around with it yet.

**Comment by pgrosu** [comment]
@chemistry-sourabh That does not seem to work.  Here is what I type:

```
(ims) [root@rhel-tgt-base ims]# bmi import bmi_infra bmi-test-image
snapshot Already Exists
(ims) [root@rhel-tgt-base ims]# bmi db ls | grep bmi-test
(ims) [root@rhel-tgt-base ims]#
(ims) [root@rhel-tgt-base ims]# rbd ls | grep bmi-test
bmi-test-image
(ims) [root@rhel-tgt-base ims]# 
```

What exactly should I be removing?  I'm writing up the tests and this is holding me up.

Thanks,
~p

**Comment by chemistry-sourabh** [comment]
You should do a "rbd ls -l", you should see the snapshot then. Then use "rbd snap rm" to remove it. If you still have issues ping me on slack

**Comment by pgrosu** [comment]
@chemistry-sourabh I guess I was not patient enough.  The `rbd ls -l` command took 30 seconds to return back results, but now the device is busy when I try to remove the snapshot.  How do ensure that we flush out the device, or is there another way to remove it?  Below is the error:

```
]# rbd ls -l | grep bmi
bmi-test-image                               1024k                                                                   2
bmi-test-image@snapshot                      1024k                                                                   2 yes
bmi_ci.img                                  10240M                                                                   2
centos_6.7_minimal@bmi-snapshot             10240M                                                                   2 yes
(ims) [root@rhel-tgt-base ims]# rbd snap rm bmi-test-image@snapshot
rbd: snapshot 'snapshot' is protected from removal.
2017-06-13 10:02:12.795767 7f237e8357c0 -1 librbd: removing snapshot from header failed: (16) Device or resource busy
```

Thanks,
~p

**Comment by pgrosu** [comment]
@chemistry-sourabh I remember Naved mentioning something about unprotecting it, so I ran the following, and it worked :)

```
# rbd snap unprotect bmi-test-image@snapshot
# rbd snap rm bmi-test-image@snapshot
# bmi import bmi_infra bmi-test-image
Success
```

Here is the documentation I followed: 

http://docs.ceph.com/docs/hammer/rbd/rbd-snapshot/

~p

**Comment by chemistry-sourabh** [comment]
Those are the right steps

**Comment by pgrosu** [comment]
Thanks @chemistry-sourabh!

**Comment by pgrosu** [comment]
@chemistry-sourabh How about this scenario?

```
# bmi db ls
+----+-------------------+-----------+-------+--------+----------+---------------+
| Id |        Name       |  Project  |  Ceph | Public | Snapshot |     Parent    |
+----+-------------------+-----------+-------+--------+----------+---------------+
| 1  |     naved-img1    | bmi_infra | 5img1 | False  |  False   |               |
| 3  | centos7-radostyle | bmi_infra | 5img3 | False  |  False   |               |
| 4  |   centos7-naved   | bmi_infra | 5img4 | False  |  False   |               |
| 6  |    test-snaphot   | bmi_infra | 5img6 | False  |   True   | centos7-naved |
| 7  |    L7r-Tt3Xb0Lg   | bmi_infra | 5img7 | False  |  False   |               |
+----+-------------------+-----------+-------+--------+----------+---------------+
# bmi rm bmi_infra L7r-Tt3Xb0Lg
192.168.122.218 - - [15/Jun/2017 22:22:25] "DELETE /remove_image/ HTTP/1.1" 500 -
L7r-Tt3Xb0Lg is Busy (Could have Clones)
#
```

**Comment by chemistry-sourabh** [comment]
Do a "rbd ls -l" and check if the image has any clones which are not flattened. Also check if this image is mounted by iscsi.

**Comment by pgrosu** [comment]
Almost there, the database is still inconsistent:

```
# rbd ls -l | grep 5img7
5img7                                        1024k                                              2
5img7@snapshot                               1024k                                              2 yes
5img8                                        1024k boot-disk-prototype/5img7@snapshot           2
# rbd snap unprotect 5img7@snapshot
2017-06-15 23:01:40.281148 7f232cecf7c0 -1 librbd: snap_unprotect: can't unprotect; at least 1 child(ren) in pool boot-disk-prototype
rbd: unprotecting snap failed: (16) Device or resource busy
(ims) [root@rhel-tgt-base ims]# rbd children 5img7@snapshot
boot-disk-prototype/5img8
# rbd flatten 5img8
Image flatten: 100% complete...done.
# rbd snap unprotect 5img7@snapshot
# rbd snap rm 5img7@snapshot
# rbd rm 5img7
Removing image: 100% complete...done.
# bmi rm bmi_infra L7r-Tt3Xb0Lg
192.168.122.218 - - [15/Jun/2017 23:10:43] "DELETE /remove_image/ HTTP/1.1" 404 -
L7r-Tt3Xb0Lg Not Found
#
```

**Comment by chemistry-sourabh** [comment]
@pgrosu Delete 5img8 if nobody is using it. Also recreate 5img7@snapshot and try deleting again

**Comment by apoorvemohan** [comment]
which BMI deployment is this?

**Comment by pgrosu** [comment]
@chemistry-sourabh Nope doesn't like it:

```
# rbd rm 5img8
2017-06-15 23:32:39.059226 7f4df063b7c0 -1 librbd: image has watchers - not removing
Removing image: 0% complete...failed.
rbd: error: image still has watchers
This means the image is still open or the client using it crashed. Try again after closing/unmapping it or waiting 30s for the crashed client to timeout.
#
# rbd snap create 5img7@snapshot
2017-06-15 23:35:49.341760 7f9a007fd7c0 -1 librbd::ImageCtx: error finding header: (2) No such file or directory
rbd: error opening image 5img7: (2) No such file or directory
#
```

@apoorvemohan This is bmi-dev at Northeastern.

**Comment by pgrosu** [comment]
Some parts works, and some didn't but still inconsistent:

```
# rbd  create 5img7 --size 1 --image-format 2
# rbd snap create 5img7@snapshot
# rbd snap rm 5img7@snapshot
# rbd rm 5img7
Removing image: 100% complete...done.
# rbd  create 5img7 --size 1 --image-format 2
# bmi rm bmi_infra L7r-Tt3Xb0Lg
192.168.122.218 - - [15/Jun/2017 23:42:39] "DELETE /remove_image/ HTTP/1.1" 404 -
snapshot Not Found
# rbd snap create 5img7@snapshot
# bmi rm bmi_infra L7r-Tt3Xb0Lg
192.168.122.218 - - [15/Jun/2017 23:43:14] "DELETE /remove_image/ HTTP/1.1" 500 -
unsupported serialized class: rbd.InvalidArgument
#
```

**Comment by apoorvemohan** [comment]
I am sure @naved001 must have informed you about this but just to re-iterate for safety - BMI-dev and BMI-prod share the same CEPH. So before removing any images make sure that you are not deleting the ones from the prod (i.e. do not remove any images that you have not created yourself).

**Comment by chemistry-sourabh** [comment]
@pgrosu I didn't see earlier that you deleted your image in ceph. So you need yo do a "bmi db rm" to remove it from db.

Regarding 5img8, I am guessing that it is mounted my iscsi. Do a "sudo tgt-admin -s" to find out

**Comment by pgrosu** [comment]
@apoorvemohan It was created by me earlier, so it was mine as it showed up in my `bmi db ls`.  Where is bmi-prod?  I was only aware of bmi-dev.  Then we definitely need to halt the release of 0.5 as we cannot have dev and production share the same Ceph, plus it is starting to fail with acceptance tests composed of randomized image names.  How do you guys then know who created what?  I always work of the `bmi db ls` information.

@chemistry-sourabh I deleted it earlier to bring the database to consistency as it was showing up created by one of my tests, and the last one that was holding me up was 5img7.  I just performed `tgt-admin --delete 5img8` and `bmi db rm bmi_infra L7r-Tt3Xb0Lg` successfully.

**Comment by apoorvemohan** [comment]
> Where is bmi-prod? 
   It's in Kaizen.

> How do you guys then know who created what?
    A unique identifier is associated with each BMI deployment that helps to differentiate between images if multiple BMI deployments share the same data store. The CEPH column seen as "bmi db ls" output is the unique name given to each image that is created. 

> Then we definitely need to halt the release of 0.5 as we cannot have dev and production share the same Ceph.
   I see no reason why you want to halt 0.5 release for the reason that the two environments share the same data store. Initially we only had single physical CEPH cluster. To test if BMI threw any error with real CEPH, we had to share a single datastore with both BMI-dev and BMI-prod. 

  I'm not sure why you guys are using BMI-dev to test your code. Weren't you guys setting up a new BMI-dev in PRB cluster? Also, why are you not using @sirushtim's scripts to setup a development environment on Kaizen VM's? 

> ...plus it is starting to fail with acceptance tests composed of randomized image names
    Can you specify the steps that resulted in error?

**Comment by apoorvemohan** [comment]
> Where is bmi-prod? 
   It's in Kaizen.

> How do you guys then know who created what?
    A unique identifier is associated with each BMI deployment that helps to differentiate between images if multiple BMI deployments share the same data store. The CEPH column seen as "bmi db ls" output is the unique name given to each image that is created. 

> Then we definitely need to halt the release of 0.5 as we cannot have dev and production share the same Ceph.
   I see no reason why you want to halt 0.5 release for the reason that the two environments share the same data store. Initially, we only had single physical CEPH cluster. To test if BMI threw any error with real CEPH, we had to share a single datastore with both BMI-dev and BMI-prod. 

  I'm not sure why you guys are using BMI-dev to test your code. Weren't you guys setting up a new BMI-dev in PRB cluster? Also, why are you not using @sirushtim's scripts to setup a development environment on Kaizen VM's? 

> ...plus it is starting to fail with acceptance tests composed of randomised image names
    Is your acceptance test using BMI command line to perform operations or RBD command line? Can you specify the steps that resulted in the error?

**Comment by naved001** [comment]
> . Weren't you guys setting up a new BMI-dev in PRB cluster?

@apoorvemohan I still haven't had the time to set it up yet. 

> Where is bmi-prod? I was only aware of bmi-dev. Then we definitely need to halt the release of 0.5 as we cannot have dev and production share the same Ceph,

@pgrosu I had explained how the Neu environment works in detail, and how we carved out a portion of the production environment to do our testing. Which is why I had been saying to be extra careful with HIL and ceph. Even yesterday I insisted that I didn't want to stress test because I did not want to make ceph busy. 

> plus it is starting to fail with acceptance tests composed of randomised image names.

As I have said before, it would great if you share your tests on github so others can see if there are some issues that we may have missed out on in our sessions.

**Comment by pgrosu** [comment]
@naved001 Yes, but as I understood from the blackboard diagram you drew for Dan (@djfinn14) and me, the NEU setup is only for those 3 nodes, which are just development (bmi-dev).  I didn't see something in the diagram for BMI production, only that the rest of the nodes are for OpenStack.  Let's sit down, and draw the storage and bmi-prod/dev operational specifications in order to put them correctly into the BMI User Guide.  Then we can come up with proposed scenarios of how to ensure the production and development environments are fully isolated from each other.  How does that sound to you?

@apoorvemohan Do you have specifications on how the UUID is used in the code and in Ceph?  You mean there is a Kaizen VM that is the BMI-Prod?  @naved001 Is Sirushti's script turn-key to the HIL setup on the PRB cluster?  I thought that you, Dan and me needed to work on Ceph and HIL configurations to make it operational.  The proper approach is to have a development Ceph storage instance that matches closely operational specifications of production, except for things such as capacity.  Testing on production using development instances can cause issues with where the data is coming from, especially as we do not have authentication.

I am putting a presentation together on the the BMI acceptance testing implementation, and the scenarios that I am running.  Once that is complete I will share it so we can all discuss it together.  Before 0.5 can be released we need step back and be completely sure we are all on the same page - on infrastructure, tests, expected supported scenarios, etc. - otherwise we will end up with unexpected surprises that will hold up our roadmap.  In the meantime, here is a test scenario I extracted out my framework that is causing issues with 0.5:

```
# ######### Test 1 ######### 
#
# rbd create L7r-Tt3Xb0Lg --size 1 --image-format 2
# rbd ls | grep L7r-Tt3Xb0Lg
L7r-Tt3Xb0Lg
# bmi import bmi_infra L7r-Tt3Xb0Lg
L7r-Tt3Xb0Lg Already Exists
# bmi db ls | grep L7r-Tt3Xb0Lg
| 7  |    L7r-Tt3Xb0Lg   | bmi_infra | 5img7 | False  |  False   |               |
# bmi rm bmi_infra L7r-Tt3Xb0Lg
192.168.122.218 - - [16/Jun/2017 09:54:56] "DELETE /remove_image/ HTTP/1.1" 500 -
unsupported serialized class: rbd.InvalidArgument
# rbd ls | grep 5img7
5img7
#
# ######### Test 2 ######### 
#
# rbd create L7r-Tt3Xb0Lz --size 1 --image-format 2
# bmi import bmi_infra L7r-Tt3Xb0Lz
# rbd create L7r-Tt3Xb0Lz --size 1 --image-format 2
# bmi import bmi_infra L7r-Tt3Xb0Lz
L7r-Tt3Xb0Lz Already Exists
# bmi db ls | grep L7r-Tt3Xb0Lz
| 8  |    L7r-Tt3Xb0Lz   | bmi_infra | 5img8 | False  |  False   |               |
#
# ######### Test 3 ######### 
#
# bmi db ls | grep L7r-Tt3Xb0Ly
# rbd create L7r-Tt3Xb0Ly --size 1 --image-format 2
# bmi db ls | grep L7r-Tt3Xb0Ly
# bmi import bmi_infra L7r-Tt3Xb0Ly
L7r-Tt3Xb0Ly Already Exists
# bmi db ls | grep L7r-Tt3Xb0Ly
| 9  |    L7r-Tt3Xb0Ly   | bmi_infra | 5img9 | False  |  False   |               |
#
```

Thanks,
Paul

**Comment by naved001** [comment]
```
# rbd ls | grep L7r-Tt3Xb0Lg
L7r-Tt3Xb0Lg
# bmi import bmi_infra L7r-Tt3Xb0Lg
L7r-Tt3Xb0Lg Already Exists
# bmi db ls | grep L7r-Tt3Xb0Lg
| 7  |    L7r-Tt3Xb0Lg   | bmi_infra | 5img7 | False  |  False   |               |
# bmi rm bmi_infra L7r-Tt3Xb0Lg
192.168.122.218 - - [16/Jun/2017 09:54:56] "DELETE /remove_image/ HTTP/1.1" 500 -
unsupported serialized class: rbd.InvalidArgument
# rbd ls | grep 5img7
5img7
```

That is concerning.

**Comment by apoorvemohan** [comment]
@pgrosu Can you briefly explain the motive of each test mentioned above?

**Comment by pgrosu** [comment]
@apoorvemohan These match how my end-to-end acceptance tests start, with some additional informational details.  Let me complete the presentation so that the whole picture becomes clear.

**Comment by apoorvemohan** [comment]
@pgrosu I want to understand the intent of the 3 test cases mentioned above. I am trying to figure out why you are getting this error.

**Comment by pgrosu** [comment]
@apoorvemohan Thanks for helping us out :) The intent is to create an image, and then to import it into BMI.  That is near the start of a multistep end-to-end acceptance test - as I have written it  - which also got agreement from both Dan and Naved.

**Comment by chemistry-sourabh** [comment]
@pgrosu are your issues solved?

**Comment by pgrosu** [comment]
@chemistry-sourabh Not quite.  So I'm not sure if you saw my email from Friday, but basically @naved001 has a small homework problem to solve of trying to figure out why those tests are failing without erasing anything.  It is probably very simple to solve - if we look carefully at the name, and how the applications processes them - but the three larger issues we need to solve afterwards are as follows:

 * How do we ensure BMI and Ceph reach consistency at startup?
 * What is a proper cleanup (maybe rollback) plan to reach consistency?
 * Proposal of a plan to have two separate Ceph storage locations, since now we use only one for both BMI-prod and BMI-dev.

One feature that cold help us is to start to add meta-data (via `rbd image-meta`) - as key-value pairs - which can help us with part of the cleanup.  The key is that we want our states (BMI/Ceph) and control flow among them to become predictable and recoverable.  If we determine edge-cases that are better covered in a future version, we should discover them now through our testing, and specify them in our release notes - otherwise we will be expected to support them, and we have a lot to do this summer.  

The last item is that we need to keep our development and production Ceph in separate locations, and not just as separate pools (which @ravisantoshgudimetla mentioned to me on Friday).  We could still be impacted on performance among pools of which there are several (check via `ceph-bmi-dev df`).  We should be able to stress-test a system for user-acceptance testing, which @naved001 recommended I not try too much on the current one.  For this last part, I'll sit down with @naved001 and get a complete layout of all our current infrastructure in order to come up with proposed design for isolated Ceph storage implementations - which we will afterwards open it up to the whole team for discussion.   

Basically we need to make sure we discover and manage against unexpected surprises before releasing any new BMI version.

Hope these are mutually agreeable,
Paul

**Comment by chemistry-sourabh** [comment]
@pgrosu if you can raise a PR for your tests. I can look at them.

**Comment by pgrosu** [comment]
@chemistry-sourabh Thank you - and appreciate the help - but this is now an individual problem for just @naved001 to tackle.  I am curious what he discovers.

**Comment by pgrosu** [comment]
Okay, here's another one that I can repeat - anyone is welcome to chip in :)  How should we deal with this situation?  Notice the uniqueness violation.  I did not clean up the database or Ceph on bmi-dev, in case you guys want to look at it:

```
(ims) [root@rhel-tgt-base ims]# bmi db ls
+----+-------------------+-----------+-------+--------+----------+---------------+
| Id |        Name       |  Project  |  Ceph | Public | Snapshot |     Parent    |
+----+-------------------+-----------+-------+--------+----------+---------------+
| 1  |     naved-img1    | bmi_infra | 5img1 | False  |  False   |               |
| 3  | centos7-radostyle | bmi_infra | 5img3 | False  |  False   |               |
| 4  |   centos7-naved   | bmi_infra | 5img4 | False  |  False   |               |
| 6  |    test-snaphot   | bmi_infra | 5img6 | False  |   True   | centos7-naved |
+----+-------------------+-----------+-------+--------+----------+---------------+
(ims) [root@rhel-tgt-base ims]# rbd ls | grep _0VyHycs0j
(ims) [root@rhel-tgt-base ims]# rbd create _0VyHycs0j --size 1 --image-format 2
(ims) [root@rhel-tgt-base ims]# rbd ls | grep _0VyHycs0j
_0VyHycs0j
(ims) [root@rhel-tgt-base ims]# bmi import bmi_infra _0VyHycs0j
Success
(ims) [root@rhel-tgt-base ims]# bmi db ls
+----+-------------------+-----------+-------+--------+----------+---------------+
| Id |        Name       |  Project  |  Ceph | Public | Snapshot |     Parent    |
+----+-------------------+-----------+-------+--------+----------+---------------+
| 1  |     naved-img1    | bmi_infra | 5img1 | False  |  False   |               |
| 3  | centos7-radostyle | bmi_infra | 5img3 | False  |  False   |               |
| 4  |   centos7-naved   | bmi_infra | 5img4 | False  |  False   |               |
| 6  |    test-snaphot   | bmi_infra | 5img6 | False  |   True   | centos7-naved |
| 7  |     _0VyHycs0j    | bmi_infra | 5img7 | False  |  False   |               |
+----+-------------------+-----------+-------+--------+----------+---------------+
(ims) [root@rhel-tgt-base ims]# bmi pro bmi_infra cisco-05 _0VyHycs0j bmi-provision-dev enp130s0f0
192.168.122.218 - - [30/Jun/2017 15:05:28] "PUT /provision/ HTTP/1.1" 200 -
Success
(ims) [root@rhel-tgt-base ims]# bmi db ls
+----+-------------------+-----------+-------+--------+----------+---------------+
| Id |        Name       |  Project  |  Ceph | Public | Snapshot |     Parent    |
+----+-------------------+-----------+-------+--------+----------+---------------+
| 1  |     naved-img1    | bmi_infra | 5img1 | False  |  False   |               |
| 3  | centos7-radostyle | bmi_infra | 5img3 | False  |  False   |               |
| 4  |   centos7-naved   | bmi_infra | 5img4 | False  |  False   |               |
| 6  |    test-snaphot   | bmi_infra | 5img6 | False  |   True   | centos7-naved |
| 7  |     _0VyHycs0j    | bmi_infra | 5img7 | False  |  False   |               |
| 8  |      cisco-05     | bmi_infra | 5img8 | False  |  False   |   _0VyHycs0j  |
+----+-------------------+-----------+-------+--------+----------+---------------+
(ims) [root@rhel-tgt-base ims]# bmi snap create bmi_infra cisco-05 _0VyHycs0j
192.168.122.218 - - [30/Jun/2017 15:05:53] "PUT /create_snapshot/ HTTP/1.1" 500 -
ORM related error = (sqlite3.IntegrityError) columns project_id, name are not unique
(ims) [root@rhel-tgt-base ims]# bmi db ls
+----+-------------------+-----------+-------+--------+----------+---------------+
| Id |        Name       |  Project  |  Ceph | Public | Snapshot |     Parent    |
+----+-------------------+-----------+-------+--------+----------+---------------+
| 1  |     naved-img1    | bmi_infra | 5img1 | False  |  False   |               |
| 3  | centos7-radostyle | bmi_infra | 5img3 | False  |  False   |               |
| 4  |   centos7-naved   | bmi_infra | 5img4 | False  |  False   |               |
| 6  |    test-snaphot   | bmi_infra | 5img6 | False  |   True   | centos7-naved |
| 7  |     _0VyHycs0j    | bmi_infra | 5img7 | False  |  False   |               |
| 8  |      cisco-05     | bmi_infra | 5img8 | False  |  False   |   _0VyHycs0j  |
+----+-------------------+-----------+-------+--------+----------+---------------+
(ims) [root@rhel-tgt-base ims]#
```

Thanks,
~p

**Comment by pgrosu** [comment]
@naved001 noticed that the snapshot name is the same as the import name - and will troubleshoot.

**Comment by naved001** [comment]
Also, please use [github gists](https://gist.github.com/) instead of posting long comments. 

>  and will troubleshoot.

What will I troubleshoot?

**Comment by pgrosu** [comment]
Agreed @naved001.

**Comment by pgrosu** [comment]
I already troubleshooted - all working now.

**Comment by chemistry-sourabh** [comment]
Sorry for the late reply.

@pgrosu this is not a bug. In BMIs point of view every snapshot is also an image as every snapshot is stored as an image in ceph. Hope this helps.

**Comment by pgrosu** [comment]
@chemistry-sourabh  No worries, and thanks for the clarification :)  We went through a session last week, in which @naved001 guided us behind the philosophy of BMI's `import` operation, and the inheritance via snapshots.  I integrated that in chapter 3 of the BMI User's Guide, which I attached here:

[BMI User's Guide.pdf](https://github.com/CCI-MOC/ims/files/1129627/BMI.User.s.Guide.pdf)

Thanks,
~p

**Comment by pgrosu** [comment]
So I completed 100 rounds of the following scenario on our development environment, and was not able to replicate this issue - maybe we can close it:

```
  Scenario: Importing/Removing Image, DB/Ceph consistency

     Given RBD will create an image
        | image_name     |
        | bmi-test-image |

     When BMI will import an image
        | image_name     | project_name  |
        | bmi-test-image | bmi_infra     |

     And BMI will remove an image
        | image_name     | project_name  |
        | bmi-test-image | bmi_infra     |

     Then RBD will confirm the removed image's clone
        | image_name     | project_name  |
        | bmi-test-image | bmi_infra     |

     And RBD will remove the created image
        | image_name     |
        | bmi-test-image |
```

What do you think?

Thanks,
`p
